File: ios/Classes/AppmetricaSdkPlugin.m
// Copyright 2019 EM ALL iT Studio. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "AppmetricaSdkPlugin.h"
#import <YandexMobileMetrica/YandexMobileMetrica.h>
#import <YandexMobileMetrica/YMMProfileAttribute.h>

@implementation AppmetricaSdkPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"emallstudio.com/appmetrica_sdk"
            binaryMessenger:[registrar messenger]];
  AppmetricaSdkPlugin* instance = [[AppmetricaSdkPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"activate" isEqualToString:call.method]) {
      [self handleActivate:call result:result];
  } else if ([@"reportEvent" isEqualToString:call.method]) {
      [self handleReportEvent:call result:result];
  } else if ([@"reportUserProfileCustomString" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomString:call result:result];
  } else if ([@"reportUserProfileCustomNumber" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomNumber:call result:result];
  } else if ([@"reportUserProfileCustomBoolean" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomBoolean:call result:result];
  } else if ([@"reportUserProfileCustomCounter" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomCounter:call result:result];
  } else if ([@"reportUserProfileUserName" isEqualToString:call.method]) {
      [self handleReportUserProfileUserName:call result:result];
  } else if ([@"reportUserProfileNotificationsEnabled" isEqualToString:call.method]) {
      [self handleReportUserProfileNotificationsEnabled:call result:result];
  } else if ([@"setStatisticsSending" isEqualToString:call.method]) {
      [self handleSetStatisticsSending:call result:result];
  } else if ([@"getLibraryVersion" isEqualToString:call.method]) {
      [self handleGetLibraryVersion:call result:result];
  } else if ([@"setUserProfileID" isEqualToString:call.method]) {
      [self handleSetUserProfileID:call result:result];
  } else if ([@"sendEventsBuffer" isEqualToString:call.method]) {
      [self handleSendEventsBuffer:call result:result];
  } else if ([@"reportReferralUrl" isEqualToString:call.method]) {
      [self handleReportReferralUrl:call result:result];
  } else if ([@"reportShowProductDetailsEvent" isEqualToString:call.method]) {
      [self handleReportShowProductDetailsEvent:call result:result];
  } else if ([@"reportAddToCartEvent" isEqualToString:call.method]) {
      [self handleReportAddToCartEvent:call result:result];
  } else if ([@"reportRemoveFromCartEvent" isEqualToString:call.method]) {
      [self handleReportRemoveFromCartEvent:call result:result];
  } else if ([@"reportBeginCheckoutEvent" isEqualToString:call.method]) {
      [self handleReportBeginCheckoutEvent:call result:result];
  } else if ([@"reportPurchaseEvent" isEqualToString:call.method]) {
      [self handleReportPurchaseEvent:call result:result];
  } else {
      result(FlutterMethodNotImplemented);
  }
}

- (FlutterError* _Nullable)getFlutterError:(NSError*)error {
  if (error == nil) return nil;

  return [FlutterError errorWithCode:[NSString stringWithFormat:@"Error %ld", (long)error.code]
                             message:error.domain
                             details:error.localizedDescription];
}

- (void)handleActivate:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* apiKey = call.arguments[@"apiKey"];
    NSInteger sessionTimeout = [call.arguments[@"sessionTimeout"] integerValue];
    BOOL locationTracking = [call.arguments[@"locationTracking"] boolValue];
    BOOL statisticsSending = [call.arguments[@"statisticsSending"] boolValue];
    BOOL crashReporting = [call.arguments[@"crashReporting"] boolValue];
    NSInteger maxReportsInDatabaseCount = [call.arguments[@"maxReportsInDatabaseCount"] integerValue];
    // Creating an extended library configuration.
    YMMYandexMetricaConfiguration *configuration = [[YMMYandexMetricaConfiguration alloc] initWithApiKey:apiKey];
    // Setting up the configuration.
    configuration.logs = YES;
    configuration.sessionTimeout = sessionTimeout;
    configuration.locationTracking = locationTracking;
    configuration.statisticsSending = statisticsSending;
    configuration.crashReporting = crashReporting;
    // maxReportsInDatabaseCount is not supported yet. Android only feature.
    //configuration.maxReportsInDatabaseCount = maxReportsInDatabaseCount;
    // Initializing the AppMetrica SDK.
    [YMMYandexMetrica activateWithConfiguration:configuration];
    result(nil);
}

- (void)handleReportEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* name = call.arguments[@"name"];

    if (![call.arguments[@"attributes"] isEqual:[NSNull null]]) {
        NSDictionary* attributes = call.arguments[@"attributes"];
        [YMMYandexMetrica reportEvent:name
            parameters:attributes
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }
        ];
    } else {
        [YMMYandexMetrica reportEvent:name
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }
        ];
    }

    result(nil);
}


- (void)handleReportUserProfileCustomString:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"value"] isEqual:[NSNull null]]) {
        NSString* value = call.arguments[@"value"];
        [userProfile apply:[[YMMProfileAttribute customString:key] withValue:value]];
    } else {
        [userProfile apply:[[YMMProfileAttribute customString:key] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileCustomNumber:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"value"] isEqual:[NSNull null]]) {
        double value = [call.arguments[@"value"] doubleValue];
        [userProfile apply:[[YMMProfileAttribute customNumber:key] withValue:value]];
    } else {
        [userProfile apply:[[YMMProfileAttribute customNumber:key] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileCustomBoolean:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"value"] isEqual:[NSNull null]]) {
        BOOL value = [call.arguments[@"value"] boolValue];
        [userProfile apply:[[YMMProfileAttribute customBool:key] withValue:value]];
    } else {
        [userProfile apply:[[YMMProfileAttribute customBool:key] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileCustomCounter:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];
    double value = [call.arguments[@"delta"] doubleValue];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    [userProfile apply:[[YMMProfileAttribute customCounter:key] withDelta:value]];
    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileUserName:(FlutterMethodCall*)call result:(FlutterResult)result {
    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"userName"] isEqual:[NSNull null]]) {
        NSString* userName = call.arguments[@"userName"];
        [userProfile apply:[[YMMProfileAttribute name] withValue:userName]];
    } else {
        [userProfile apply:[[YMMProfileAttribute name] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileNotificationsEnabled:(FlutterMethodCall*)call result:(FlutterResult)result {
    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"notificationsEnabled"] isEqual:[NSNull null]]) {
        BOOL notificationsEnabled = [call.arguments[@"notificationsEnabled"] boolValue];
        [userProfile apply:[[YMMProfileAttribute notificationsEnabled] withValue:notificationsEnabled]];
    } else {
        [userProfile apply:[[YMMProfileAttribute notificationsEnabled] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleSetStatisticsSending:(FlutterMethodCall*)call result:(FlutterResult)result {
    BOOL statisticsSending = [call.arguments[@"statisticsSending"] boolValue];

    [YMMYandexMetrica setStatisticsSending:statisticsSending];

    result(nil);
}

- (void)handleGetLibraryVersion:(FlutterMethodCall*)call result:(FlutterResult)result {
    result([YMMYandexMetrica libraryVersion]);
}

- (void)handleSetUserProfileID:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* userProfileID = call.arguments[@"userProfileID"];

    [YMMYandexMetrica setUserProfileID:userProfileID];

    result(nil);
}

- (void)handleSendEventsBuffer:(FlutterMethodCall*)call result:(FlutterResult)result {
    [YMMYandexMetrica sendEventsBuffer];

    result(nil);
}

- (void)handleReportReferralUrl:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* referral = call.arguments[@"referral"];
    NSURL *url = [[NSURL alloc] initWithString:referral];
    
    [YMMYandexMetrica reportReferralUrl:url];

    result(nil);
}


- (void)handleReportShowProductDetailsEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
    if (![call.arguments[@"attributes"] isEqual:[NSNull null]]) {
        printf("report product details");
        NSDictionary* attributes = call.arguments[@"attributes"];
        NSString* productName = attributes[@"product"];
        NSString* productCode = attributes[@"productCode"];
        double priceValue = [attributes[@"price"] doubleValue];
        YMMECommerceAmount *actualFiat =
        [[YMMECommerceAmount alloc] initWithUnit:@"RUB" value:[[NSDecimalNumber alloc] initWithDouble:priceValue]];
        YMMECommercePrice *actualPrice = [[YMMECommercePrice alloc] initWithFiat:actualFiat];
        YMMECommerceProduct *product = [[YMMECommerceProduct alloc] initWithSKU:productCode
                                                                           name:productName
                                                             categoryComponents:nil
                                                                        payload:nil
                                                                    actualPrice:actualPrice
                                                                  originalPrice:nil
                                                                     promoCodes:nil];
        [YMMYandexMetrica reportECommerce:[YMMECommerce showProductDetailsEventWithProduct:product referrer:nil] onFailure:^(NSError *error) {
            result([self getFlutterError:error]);
         }];
    }

    result(nil);
}

- (void)handleReportAddToCartEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
    printf("report add to cart");
    if (![call.arguments[@"attributes"] isEqual:[NSNull null]]) {
        NSDictionary* attributes = call.arguments[@"attributes"];
        YMMECommerceCartItem *addedItems = [self createCartItem: attributes];
        [YMMYandexMetrica reportECommerce:[YMMECommerce addCartItemEventWithItem:addedItems] onFailure:^(NSError *error) {
            result([self getFlutterError:error]);
         }];
        
    }
    result(nil);
}

- (void)handleReportRemoveFromCartEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
    printf("report remove from cart");
    if (![call.arguments[@"attributes"] isEqual:[NSNull null]]) {
        NSDictionary* attributes = call.arguments[@"attributes"];
        YMMECommerceCartItem *removedItems = [self createCartItem: attributes];
        
        [YMMYandexMetrica reportECommerce:[YMMECommerce removeCartItemEventWithItem:removedItems] onFailure:^(NSError *error) {
            result([self getFlutterError:error]);
         }];
        
    }
    result(nil);
}

- (void)handleReportBeginCheckoutEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
    printf("report remove from cart");
    if (![call.arguments[@"cartItems"] isEqual:[NSNull null]]) {
        NSDictionary* cartItems = call.arguments[@"cartItems"];
        NSString* cartId = call.arguments[@"cartId"];
        NSLog(@"%@", cartId);
        NSMutableArray <YMMECommerceCartItem *> *cartItemsArray = [self createCartItems:cartItems];
        for (YMMECommerceCartItem* v in cartItemsArray){
            NSLog(@"cartItem added");
            NSLog(@"%@", v.product.name);
        }
        
        YMMECommerceOrder *order = [[YMMECommerceOrder alloc] initWithIdentifier:cartId
                                                                       cartItems:cartItemsArray
                                                                         payload:nil];
        
        [YMMYandexMetrica reportECommerce:[YMMECommerce beginCheckoutEventWithOrder:order] onFailure:^(NSError *error) {
            result([self getFlutterError:error]);
         }];
    }
    result(nil);
}


- (void)handleReportPurchaseEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
    printf("report purchase");
    if (![call.arguments[@"cartItems"] isEqual:[NSNull null]]) {
        NSDictionary* cartItems = call.arguments[@"cartItems"];
        NSString* cartId = call.arguments[@"cartId"];
        NSLog(@"%@", cartId);
        NSMutableArray <YMMECommerceCartItem *> *cartItemsArray = [self createCartItems:cartItems];
        for (YMMECommerceCartItem* v in cartItemsArray){
            NSLog(@"cartItem added");
            NSLog(@"%@", v.product.name);
        }
        
        YMMECommerceOrder *order = [[YMMECommerceOrder alloc] initWithIdentifier:cartId
                                                                       cartItems:cartItemsArray
                                                                         payload:nil];
        
        [YMMYandexMetrica reportECommerce:[YMMECommerce purchaseEventWithOrder:order] onFailure:^(NSError *error) {
            result([self getFlutterError:error]);
         }];
    }
    result(nil);
}

- (NSMutableArray <YMMECommerceCartItem *>*)createCartItems:(NSDictionary*) items{
    NSMutableArray <YMMECommerceCartItem *> *cartProducts = [[NSMutableArray alloc] initWithCapacity:items.count];
    for (NSDictionary* item in items){
        YMMECommerceCartItem* cartItem = [self createCartItem:item[@"attributes"]];
        [cartProducts addObject:cartItem];
    }
    return cartProducts;
}

- (YMMECommerceCartItem*)createCartItem:(NSDictionary*) attributes{
    NSString* productName = attributes[@"product"];
    NSString* productCode = attributes[@"productCode"];
    double priceValue = [attributes[@"price"] doubleValue];
    int quantityValue = [attributes[@"quantity"] intValue];
    YMMECommerceAmount *actualFiat =
    [[YMMECommerceAmount alloc] initWithUnit:@"RUB" value:[[NSDecimalNumber alloc] initWithDouble:priceValue]];
    YMMECommercePrice *actualPrice = [[YMMECommercePrice alloc] initWithFiat:actualFiat];
    YMMECommerceProduct *product = [[YMMECommerceProduct alloc] initWithSKU:productCode
                                                                       name:productName
                                                         categoryComponents:nil
                                                                    payload:nil
                                                                actualPrice:actualPrice
                                                              originalPrice:nil
                                                                 promoCodes:nil];
    YMMECommerceCartItem *cartItem = [[YMMECommerceCartItem alloc] initWithProduct:product quantity:[[NSDecimalNumber alloc] initWithInt:quantityValue] revenue:actualPrice referrer:nil];
    return cartItem;
}

@end

File: android/src/main/java/com/emallstudio/appmetrica_sdk/AppmetricaSdkPlugin.java
// Copyright 2019 EM ALL iT Studio. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.emallstudio.appmetrica_sdk;

import androidx.annotation.NonNull;

import android.util.Log;
import android.app.Application;
import android.content.Context;

import java.util.ArrayList;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import com.yandex.metrica.YandexMetrica;
import com.yandex.metrica.YandexMetricaConfig;
import com.yandex.metrica.ecommerce.ECommerceAmount;
import com.yandex.metrica.ecommerce.ECommerceCartItem;
import com.yandex.metrica.ecommerce.ECommerceEvent;
import com.yandex.metrica.ecommerce.ECommerceOrder;
import com.yandex.metrica.ecommerce.ECommercePrice;
import com.yandex.metrica.ecommerce.ECommerceProduct;
import com.yandex.metrica.profile.Attribute;
import com.yandex.metrica.profile.UserProfile;

/**
 * AppmetricaSdkPlugin
 */
public class AppmetricaSdkPlugin implements MethodCallHandler, FlutterPlugin {
    private static final String TAG = "AppmetricaSdkPlugin";
    private MethodChannel methodChannel;
    private Context context;
    private Application application;

    /**
     * Plugin registration for v1 embedder.
     */
    public static void registerWith(Registrar registrar) {
        final AppmetricaSdkPlugin instance = new AppmetricaSdkPlugin();
        instance.onAttachedToEngine(registrar.context(), registrar.messenger());
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        onAttachedToEngine(flutterPluginBinding.getApplicationContext(), flutterPluginBinding.getBinaryMessenger());
    }

    private void onAttachedToEngine(Context applicationContext, BinaryMessenger binaryMessenger) {
        application = (Application) applicationContext;
        context = applicationContext;
        methodChannel = new MethodChannel(binaryMessenger, "emallstudio.com/appmetrica_sdk");
        methodChannel.setMethodCallHandler(this);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        methodChannel.setMethodCallHandler(null);
        methodChannel = null;
        context = null;
        application = null;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "activate":
                handleActivate(call, result);
                break;
            case "reportEvent":
                handleReportEvent(call, result);
                break;
            case "reportShowProductDetailsEvent":
                handleReportShowProductDetailsEvent(call, result);
                break;
            case "reportAddToCartEvent":
                handleReportAddToCartEvent(call, result);
                break;
            case "reportRemoveFromCartEvent":
                handleReportRemoveFromCartEvent(call, result);
                break;
            case "reportBeginCheckoutEvent":
                handleReportBeginCheckoutEvent(call, result);
                break;
            case "reportPurchaseEvent":
                handleReportPurchaseEvent(call, result);
                break;
            case "reportUserProfileCustomString":
                handleReportUserProfileCustomString(call, result);
                break;
            case "reportUserProfileCustomNumber":
                handleReportUserProfileCustomNumber(call, result);
                break;
            case "reportUserProfileCustomBoolean":
                handleReportUserProfileCustomBoolean(call, result);
                break;
            case "reportUserProfileCustomCounter":
                handleReportUserProfileCustomCounter(call, result);
                break;
            case "reportUserProfileUserName":
                handleReportUserProfileUserName(call, result);
                break;
            case "reportUserProfileNotificationsEnabled":
                handleReportUserProfileNotificationsEnabled(call, result);
                break;
            case "setStatisticsSending":
                handleSetStatisticsSending(call, result);
                break;
            case "getLibraryVersion":
                handleGetLibraryVersion(call, result);
                break;
            case "setUserProfileID":
                handleSetUserProfileID(call, result);
                break;
            case "sendEventsBuffer":
                handleSendEventsBuffer(call, result);
                break;
            case "reportReferralUrl":
                handleReportReferralUrl(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void handleActivate(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            // Get activation parameters.
            final String apiKey = (String) arguments.get("apiKey");
            final int sessionTimeout = (int) arguments.get("sessionTimeout");
            final boolean locationTracking = (boolean) arguments.get("locationTracking");
            final boolean statisticsSending = (boolean) arguments.get("statisticsSending");
            final boolean crashReporting = (boolean) arguments.get("crashReporting");
            final int maxReportsInDatabaseCount = (int) arguments.get("maxReportsInDatabaseCount");
            // Creating an extended library configuration.
            YandexMetricaConfig config = YandexMetricaConfig.newConfigBuilder(apiKey)
                    .withLogs()
                    .withSessionTimeout(sessionTimeout)
                    .withLocationTracking(locationTracking)
                    .withStatisticsSending(statisticsSending)
                    .withCrashReporting(crashReporting)
                    .withMaxReportsInDatabaseCount(maxReportsInDatabaseCount)
                    .build();
            // Initializing the AppMetrica SDK.
            YandexMetrica.activate(context, config);
            // Automatic tracking of user activity.
            YandexMetrica.enableActivityAutoTracking(application);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error performing activation", e.getMessage(), null);
        }
        result.success(null);
    }

    private void handleReportEvent(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String eventName = (String) arguments.get("name");
            @SuppressWarnings("unchecked")
            Map<String, Object> attributes = (Map<String, Object>) arguments.get("attributes");
            if (attributes == null) {
                YandexMetrica.reportEvent(eventName);
            } else {
                YandexMetrica.reportEvent(eventName, attributes);
            }

        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing event", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportShowProductDetailsEvent(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            @SuppressWarnings("unchecked")
            Map<String, Object> attributes = (Map<String, Object>) arguments.get("attributes");

            if (attributes != null) {
                final String productName = (String) attributes.get("product");
                final String productCode = (String) attributes.get("productCode");
                final Double price = (Double) attributes.get("price");
                System.out.println("name: " + productName + " code: " + productCode + " price:" + price);
                if (productName == null || productCode == null || price == null) return;

                ECommercePrice actualPrice = new ECommercePrice(new ECommerceAmount(price, "RUB"));
                ECommerceProduct product = new ECommerceProduct(productCode)
                        .setActualPrice(actualPrice) // Optional.
                        .setName(productName); // Optional.

                ECommerceEvent showProductDetailsEvent = ECommerceEvent.showProductDetailsEvent(product, null);
                YandexMetrica.reportECommerce(showProductDetailsEvent);
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporting event", e.getMessage(), null);
        }
        result.success(null);
    }

    private void handleReportAddToCartEvent(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            @SuppressWarnings("unchecked")
            Map<String, Object> attributes = (Map<String, Object>) arguments.get("attributes");
            if (attributes != null) {
                ECommerceCartItem cartItem = produceCartItem(attributes);
                if (cartItem != null) {
                    ECommerceEvent addCartItemEvent = ECommerceEvent.addCartItemEvent(cartItem);
                    YandexMetrica.reportECommerce(addCartItemEvent);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporting event", e.getMessage(), null);
        }
        result.success(null);
    }

    private void handleReportRemoveFromCartEvent(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            @SuppressWarnings("unchecked")
            Map<String, Object> attributes = (Map<String, Object>) arguments.get("attributes");
            if (attributes != null) {
                ECommerceCartItem cartItem = produceCartItem(attributes);
                if (cartItem != null) {
                    ECommerceEvent removeCartEvent = ECommerceEvent.removeCartItemEvent(cartItem);
                    YandexMetrica.reportECommerce(removeCartEvent);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporting event", e.getMessage(), null);
        }
        result.success(null);
    }

    private void handleReportBeginCheckoutEvent(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String cartId = (String) arguments.get("cartId");
            @SuppressWarnings("unchecked")
            ArrayList<Map<String, Object>> cartItems = (ArrayList<Map<String, Object>>) arguments.get("cartItems");
            if (cartItems != null && cartId != null) {
                ArrayList<ECommerceCartItem> cartItemsList = produceECCartItems(cartItems);
                ECommerceOrder order = new ECommerceOrder(cartId, cartItemsList);
                ECommerceEvent beginCheckoutEvent = ECommerceEvent.beginCheckoutEvent(order);
                YandexMetrica.reportECommerce(beginCheckoutEvent);
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporting event", e.getMessage(), null);
        }
        result.success(null);
    }

    private void handleReportPurchaseEvent(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String cartId = (String) arguments.get("cartId");
            @SuppressWarnings("unchecked")
            ArrayList<Map<String, Object>> cartItems = (ArrayList<Map<String, Object>>) arguments.get("cartItems");
            if (cartItems != null && cartId != null) {
                ArrayList<ECommerceCartItem> cartItemsList = produceECCartItems(cartItems);
                ECommerceOrder order = new ECommerceOrder(cartId, cartItemsList);
                ECommerceEvent purchaseEvent = ECommerceEvent.purchaseEvent(order);
                YandexMetrica.reportECommerce(purchaseEvent);
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporting event", e.getMessage(), null);
        }
        result.success(null);
    }

    private ArrayList<ECommerceCartItem> produceECCartItems(ArrayList<Map<String, Object>> cartItems) {
        ArrayList<ECommerceCartItem> cartItemsList = new ArrayList<>();
        for (Map<String, Object> item : cartItems) {
            @SuppressWarnings("unchecked")
            Map<String, Object> itemAttributes = (Map<String, Object>) item.get("attributes");
            if (itemAttributes != null) {
                ECommerceCartItem cartItem = produceCartItem(itemAttributes);
                if (cartItem != null) {
                    cartItemsList.add(cartItem);
                }
            }
        }
        for (ECommerceCartItem i : cartItemsList) {
            System.out.println(i.toString());
        }
        return cartItemsList;
    }

    private ECommerceCartItem produceCartItem(Map<String, Object> itemAttributes) {
        final String productName = (String) itemAttributes.get("product");
        final String productCode = (String) itemAttributes.get("productCode");
        final Double price = (Double) itemAttributes.get("price");
        final Integer quantity = (Integer) itemAttributes.get("quantity");

        if (productName == null || productCode == null || price == null || quantity == null)
            return null;

        ECommercePrice actualPrice = new ECommercePrice(new ECommerceAmount(price, "RUB"));
        ECommerceProduct product = new ECommerceProduct(productCode)
                .setActualPrice(actualPrice) // Optional.
                .setName(productName); // Optional.
        return new ECommerceCartItem(product, actualPrice, quantity);
    }

    private void handleReportUserProfileCustomString(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            final String value = (String) arguments.get("value");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (value != null) {
                profileBuilder.apply(Attribute.customString(key).withValue(value));
            } else {
                profileBuilder.apply(Attribute.customString(key).withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom string", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileCustomNumber(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("value") != null) {
                final double value = (double) arguments.get("value");
                profileBuilder.apply(Attribute.customNumber(key).withValue(value));
            } else {
                profileBuilder.apply(Attribute.customNumber(key).withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom number", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileCustomBoolean(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("value") != null) {
                final boolean value = (boolean) arguments.get("value");
                profileBuilder.apply(Attribute.customBoolean(key).withValue(value));
            } else {
                profileBuilder.apply(Attribute.customBoolean(key).withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom boolean", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileCustomCounter(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            final double delta = (double) arguments.get("delta");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            profileBuilder.apply(Attribute.customCounter(key).withDelta(delta));
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom counter", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileUserName(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("userName") != null) {
                final String userName = (String) arguments.get("userName");
                profileBuilder.apply(Attribute.name().withValue(userName));
            } else {
                profileBuilder.apply(Attribute.name().withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile user name", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileNotificationsEnabled(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("notificationsEnabled") != null) {
                final boolean notificationsEnabled = (boolean) arguments.get("notificationsEnabled");
                profileBuilder.apply(Attribute.notificationsEnabled().withValue(notificationsEnabled));
            } else {
                profileBuilder.apply(Attribute.notificationsEnabled().withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile user name", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleSetStatisticsSending(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final boolean statisticsSending = (boolean) arguments.get("statisticsSending");
            YandexMetrica.setStatisticsSending(context, statisticsSending);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error enable sending statistics", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleGetLibraryVersion(MethodCall call, Result result) {
        try {
            result.success(YandexMetrica.getLibraryVersion());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error enable sending statistics", e.getMessage(), null);
        }
    }

    private void handleSetUserProfileID(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String userProfileID = (String) arguments.get("userProfileID");
            YandexMetrica.setUserProfileID(userProfileID);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error sets the ID of the user profile", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleSendEventsBuffer(MethodCall call, Result result) {
        try {
            YandexMetrica.sendEventsBuffer();
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error sending stored events from the buffer", e.getMessage(), null);

        }

        result.success(null);
    }

    private void handleReportReferralUrl(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String referral = (String) arguments.get("referral");
            YandexMetrica.reportReferralUrl(referral);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error sets the ID of the user profile", e.getMessage(), null);
        }

        result.success(null);
    }
}
